tictactoe/game1.py

```diff
@@ -54,46 +62,128 @@ def is_draw(self):
-            print(self)
+
+class Undo(Action):
+    def execute(self, game: Game):
+        if game.marked_tiles() != 0:
+            game.board[game.filled_tiles[-1]] = Mark._
+            game.filled_tiles.pop()
+            print(game)
         else:
             print("cant undo")
 
 
-class main:
-    filled_tiles = []
-    game = Game(filled_tiles)
-    print(game)
+class StorageDriver(abc.ABC):
+    @abc.abstractmethod
+    def store_move(self, update: Update): ...
+    @abc.abstractmethod
+    def read_moves(self) -> list[Update]: ...
+    @abc.abstractmethod
+    def undo_move(self): ...
+    def delete_moves(self): ...
+
+
+@dc.dataclass
+class StoreDB(StorageDriver):
+    def __post_init__(self):
+        self.con = sqlite3.connect("tictactoe.db")
+        self.cur = self.con.cursor()
+        self.cur.execute(
+            "CREATE TABLE IF NOT EXISTS moves(id INTEGER PRIMARY KEY, tile INTEGER)"
+        )
+        self.con.commit()
+
+    def store_move(self, update: Update):
+        self.cur.execute("INSERT INTO moves(tile) VALUES (?)", (update.tile,))
+        self.con.commit()
+
+    def read_moves(self) -> list[Update]:
+        res = self.cur.execute("SELECT tile FROM moves")
+        moves = res.fetchall()
+        print("fetchingg", moves)
+        print("listttt", [Update(move[0]) for move in moves])
+        return [Update(move[0]) for move in moves]
+
+    def undo_move(self):
+        self.cur.execute("DELETE FROM moves WHERE id = (SELECT MAX(id) FROM moves)")
+        self.con.commit()
+
+    def delete_moves(self):
+        self.cur.execute("DELETE FROM moves")
+        self.con.commit()
+
+
+@dc.dataclass
+class DBUpdate(Action):
+    sd: StoreDB
+    update: Update
+
+    def execute(self, game: Game):
+        self.sd.store_move(self.update)
+        self.update.execute(game)
+
+
+@dc.dataclass
+class DBUndo(Action):
+    sd: StoreDB
+    undo: Undo
+
+    def execute(self, game: Game):
+        self.sd.undo_move()
+        self.undo.execute(game)
+
+
+def main():
+    game = Game()
+
+    sd = StoreDB()
+    for update in sd.read_moves():
+        game.apply(update)
+
     while True:
         input1 = int(input("pick a tile "))
         if input1 == -1:
-            game.undo()
+            action = DBUndo(sd, Undo())
         else:
-            game.update_board(input1)
-
-        x_winner = game.is_winner("x")
-        o_winner = game.is_winner("o")
-        is_draw = game.is_draw()
-        if x_winner:
+            action = DBUpdate(sd, Update(tile=input1))
+        game.apply(action)
+        if game.is_winner(Mark.x):
             print("x is the winner")
+            sd.delete_moves()
             break
-        if o_winner:
+        elif game.is_winner(Mark.o):
             print("o is the winner")
+            sd.delete_moves()
             break
-        if is_draw:
+        elif game.is_draw():
             print("its a draw")
+            sd.delete_moves()
             break
```

Comment:
instead of 2 methods – `is_winner` and `is_draw` – use 1 method called `ended` which returns `(ended: bool, winner: Mark)`